File: examples/internal/delay/agent.py
"""Change Window Validator — internal/delay example.

Delivery: stream  (GET /v1/agents/{address}/intents/stream)

Validates that a change falls within an approved maintenance window.
Demonstrates a real async validation pattern: the agent checks the
window schedule and validates all required fields before responding.
The step has a 120-second deadline (`step_deadline_seconds: 120`).

Run (Terminal 1 — agent):
    export AXME_API_KEY=<your-api-key>
    export AXME_AGENT_ADDRESS=agent://<org>/<workspace>/change-window-validator
    python agent.py

Run (Terminal 2 — scenario):
    axme scenarios apply examples/internal/delay/scenario.json --watch

To observe step deadline: stop the agent before it responds.
AXME will transition the step to TIMED_OUT after 120 seconds.
"""
from __future__ import annotations

import logging
import os
import sys
from datetime import datetime, timezone
from typing import Any

from axme import AxmeClient, AxmeClientConfig

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  [%(levelname)s]  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("change-window-validator")

AXME_BASE_URL      = os.environ.get("AXME_BASE_URL", "https://api.cloud.axme.ai")
AXME_API_KEY       = os.environ.get("AXME_API_KEY", "")
AXME_AGENT_ADDRESS = os.environ.get("AXME_AGENT_ADDRESS", "change-window-validator")

if not AXME_API_KEY:
    sys.exit("AXME_API_KEY is required")

# ---------------------------------------------------------------------------
# Change window validation rules
# ---------------------------------------------------------------------------
# Approved maintenance windows: weekday evenings and weekends.
# Changes outside these windows are rejected.

_APPROVED_WINDOW_IDS = {
    "MW-2026-Q1-01", "MW-2026-Q1-02", "MW-2026-Q1-03",
    "MW-2026-Q1-04", "MW-2026-Q1-05", "MW-2026-Q1-06",
}

_ALLOWED_CHANGE_TYPES = {"config_update", "dependency_update", "rollback", "hotfix"}

_BLOCKED_ENVIRONMENTS_WITHOUT_WINDOW = {"production"}


def _validate_change_window(payload: dict[str, Any]) -> tuple[bool, str, dict[str, Any]]:
    """Return (allowed, reason, metadata)."""
    change_id    = payload.get("change_id", "?")
    environment  = payload.get("environment", "")
    change_type  = payload.get("change_type", "")
    window_id    = payload.get("window_id", "")
    scheduled_at = payload.get("scheduled_at", "")

    # Check 1 — change type allowed
    if change_type not in _ALLOWED_CHANGE_TYPES:
        return (
            False,
            f"{change_id}: change_type {change_type!r} is not in allowed set {sorted(_ALLOWED_CHANGE_TYPES)}",
            {},
        )

    # Check 2 — production changes must have an approved window
    if environment in _BLOCKED_ENVIRONMENTS_WITHOUT_WINDOW:
        if not window_id:
            return (
                False,
                f"{change_id}: production changes require an approved window_id",
                {},
            )
        if window_id not in _APPROVED_WINDOW_IDS:
            return (
                False,
                f"{change_id}: window_id {window_id!r} is not in the approved window list",
                {},
            )

    # Check 3 — scheduled_at is a valid ISO timestamp
    if scheduled_at:
        try:
            datetime.fromisoformat(scheduled_at.replace("Z", "+00:00"))
        except ValueError:
            return (
                False,
                f"{change_id}: scheduled_at {scheduled_at!r} is not a valid ISO timestamp",
                {},
            )

    window_info = {}
    if window_id and window_id in _APPROVED_WINDOW_IDS:
        window_info = {"window_id": window_id, "window_approved": True}

    return (
        True,
        f"{change_id}: change window validated for {environment!r} environment",
        window_info,
    )


def handle_intent(client: AxmeClient, delivery: dict[str, Any]) -> None:
    intent_id = delivery.get("intent_id", "")
    if not intent_id:
        log.warning("received delivery without intent_id, skipping")
        return

    log.info("received intent %s", intent_id)

    try:
        resp   = client.get_intent(intent_id)
        intent = resp.get("intent") or resp
    except Exception as exc:
        log.error("get_intent(%s) failed: %s", intent_id, exc)
        return

    status = (intent.get("lifecycle_status") or intent.get("status") or "").upper()
    if status not in {"CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"}:
        log.debug("skipping intent %s with status %s", intent_id, status)
        return

    raw_payload       = intent.get("payload") or {}
    effective_payload = raw_payload.get("parent_payload") or raw_payload

    allowed, reason, meta = _validate_change_window(effective_payload)
    log.info("window validation for %s: allowed=%s  reason=%s", intent_id, allowed, reason)

    try:
        if allowed:
            client.resume_intent(
                intent_id,
                {
                    "action":  "complete",
                    "allowed": True,
                    "reason":  reason,
                    **meta,
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        else:
            client.resume_intent(
                intent_id,
                {
                    "action":  "fail",
                    "allowed": False,
                    "reason":  reason,
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        log.info("resumed intent %s (allowed=%s)", intent_id, allowed)
    except Exception as exc:
        log.error("resume_intent(%s) failed: %s", intent_id, exc)


def main() -> None:
    client = AxmeClient(AxmeClientConfig(api_key=AXME_API_KEY, base_url=AXME_BASE_URL))
    log.info(
        "change-window-validator starting  address=%s  binding=stream",
        AXME_AGENT_ADDRESS,
    )

    try:
        for delivery in client.listen(AXME_AGENT_ADDRESS, timeout_seconds=None):
            handle_intent(client, delivery)
    except KeyboardInterrupt:
        log.info("shutting down")


if __name__ == "__main__":
    main()

File: examples/internal/delay/agent.ts
/**
 * Change Window Validator — internal/delay example (TypeScript).
 *
 * Validates that a change falls within an approved maintenance window.
 * The step has a 120-second deadline (step_deadline_seconds: 120).
 *
 * Run:
 *   export AXME_API_KEY=<your-api-key>
 *   export AXME_AGENT_ADDRESS=change-window-validator
 *   npx tsx examples/internal/delay/agent.ts
 */
import { AxmeClient } from "@axme/axme";

const AXME_BASE_URL = process.env.AXME_BASE_URL ?? "https://api.cloud.axme.ai";
const AXME_API_KEY = process.env.AXME_API_KEY ?? "";
const AXME_AGENT_ADDRESS = process.env.AXME_AGENT_ADDRESS ?? "change-window-validator";

if (!AXME_API_KEY) { console.error("AXME_API_KEY is required"); process.exit(1); }

// --- Business logic ---
const APPROVED_WINDOWS = new Set(["MW-2026-Q1-01", "MW-2026-Q1-02", "MW-2026-Q1-03", "MW-2026-Q1-04", "MW-2026-Q1-05", "MW-2026-Q1-06"]);
const ALLOWED_CHANGE_TYPES = new Set(["config_update", "dependency_update", "rollback", "hotfix"]);

function validateChangeWindow(payload: Record<string, unknown>): { allowed: boolean; reason: string; meta: Record<string, unknown> } {
  const changeId = String(payload.change_id ?? "?");
  const environment = String(payload.environment ?? "");
  const changeType = String(payload.change_type ?? "");
  const windowId = String(payload.window_id ?? "");
  const scheduledAt = String(payload.scheduled_at ?? "");

  if (!ALLOWED_CHANGE_TYPES.has(changeType)) {
    return { allowed: false, reason: `${changeId}: change_type '${changeType}' is not in allowed set ${[...ALLOWED_CHANGE_TYPES].sort()}`, meta: {} };
  }
  if (environment === "production") {
    if (!windowId) return { allowed: false, reason: `${changeId}: production changes require an approved window_id`, meta: {} };
    if (!APPROVED_WINDOWS.has(windowId)) return { allowed: false, reason: `${changeId}: window_id '${windowId}' is not in the approved window list`, meta: {} };
  }
  if (scheduledAt) {
    const d = new Date(scheduledAt);
    if (isNaN(d.getTime())) return { allowed: false, reason: `${changeId}: scheduled_at '${scheduledAt}' is not a valid ISO timestamp`, meta: {} };
  }

  const meta: Record<string, unknown> = {};
  if (windowId && APPROVED_WINDOWS.has(windowId)) Object.assign(meta, { window_id: windowId, window_approved: true });

  return { allowed: true, reason: `${changeId}: change window validated for '${environment}' environment`, meta };
}

// --- Agent loop ---
const client = new AxmeClient({ apiKey: AXME_API_KEY, baseUrl: AXME_BASE_URL });
console.log(`change-window-validator starting  address=${AXME_AGENT_ADDRESS}  binding=stream`);

try {
  for await (const delivery of client.listen(AXME_AGENT_ADDRESS)) {
    const intentId = String(delivery.intent_id ?? "");
    if (!intentId) continue;
    console.log(`received intent ${intentId}`);

    let intent: Record<string, unknown>;
    try {
      const resp = await client.getIntent(intentId);
      intent = (resp.intent ?? resp) as Record<string, unknown>;
    } catch (e) { console.error(`get_intent(${intentId}) failed:`, e); continue; }

    const status = String(intent.lifecycle_status ?? intent.status ?? "").toUpperCase();
    if (!["CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"].includes(status)) continue;

    const rawPayload = (intent.payload ?? {}) as Record<string, unknown>;
    const effectivePayload = (rawPayload.parent_payload ?? rawPayload) as Record<string, unknown>;
    const { allowed, reason, meta } = validateChangeWindow(effectivePayload);

    console.log(`window validation for ${intentId}: allowed=${allowed}  reason=${reason}`);

    try {
      if (allowed) {
        await client.resumeIntent(intentId, { action: "complete", allowed: true, reason, ...meta }, { ownerAgent: AXME_AGENT_ADDRESS });
      } else {
        await client.resumeIntent(intentId, { action: "fail", allowed: false, reason }, { ownerAgent: AXME_AGENT_ADDRESS });
      }
      console.log(`resumed intent ${intentId} (allowed=${allowed})`);
    } catch (e) { console.error(`resume_intent(${intentId}) failed:`, e); }
  }
} catch (e) {
  if ((e as Error).name !== "AbortError") throw e;
}
